[WIP] Refactor ITC code

bas-rustenburg opened the pull request:
Plan is to move stuff out of ITC.py, and to make use of experiment and instrument classes in constructing models.

Commit:
Refactoring in progress: extract instruments and experiment classes.
- Experiment and Injection class have been placed in experiments.py
- Instrument and VP-ITC class have been extracted and simtk.unit has taken the place of Units.py

Commit:
Extracted report.

Commit:
Cleaned up comments.

Commit:
Updated comments in ITC.py

Commit:
Deleted pymc models from ITC.py.
Next step is to actually use models from models.py in it's place.
Slight update to some comments.

Commit:
Remove reference to global variable experiments in ITC.analyze

Commit:
Moved analyze to report.py

Commit:
Remove redundant parentheses, declare VPITC.

Commit:
Replace commands (deprecated) by glob.

Commit:
Make name extraction more human readable.

Commit:
Update comments ITC.py

Commit:
Import os in experiments.py

Commit:
Use  simtk.unit in experiments.py

Commit:
Import numpy in experiments.py

Commit:
Fixed incompatibilities numpy float array and unit.Quantity

Commit:
Added a temporary workaround for SimTk/openmm#642.

Commit:
Missing sum feature was added in SimTk/openmm#643.

Commit:
Provide backward compatibility with simtk.unit.

Commit:
Updated analyze in report.py

Commit:
Fixed experiment.write_integrated_heats

Commit:
Replaced Units.py by simtk.unit in ITC.py

Commit:
Minor modifications in models, lots of indent fixes.

Commit:
Now using Injection class.

Commit:
Take injection heats from Injection.evolved_heat

jchodera commented on python/refactored/ITC.py:
```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/python
+
+#=============================================================================================
```
Change this to a docstring.

jchodera commented on python/refactored/ITC.py:
```diff
@@ -0,0 +1,191 @@
+#=============================================================================================
+
+import os
+from simtk import unit
```
Switch to [`pint`](http://pint.readthedocs.org/en/0.6/).

bas-rustenburg replied:
Switched to `pint` for unit. Was very easy, since the syntax is almost identical to `simtk.unit`.

jchodera commented on python/refactored/ITC.py:
```diff
@@ -0,0 +1,191 @@
+from models import RescalingStep, TwoComponentBindingModel
+
+
+def compute_statistics(x_t):
```
This may not be used anymore.  Would call it `compute_normal_statistics` or something, and add docstring.

jchodera commented on python/refactored/ITC.py:
```diff
@@ -0,0 +1,191 @@
+    from glob import glob
+    from os.path import basename, splitext
+    
+    directory = '../../data/auto-iTC-200/053014/'
```
Add [`docopt`](http://docopt.org) option parser.  Focus on command-line that permits batch automated processing of `.itc` and `.dh` files, printing out PDF and numerical reports.

jchodera commented on python/refactored/ITC.py:
```diff
@@ -0,0 +1,191 @@
+        analyze(experiment_name, experiment)
+        # Write Origin-style integrated heats.
+        filename = experiment_name + '-integrated.txt'
+        experiment.write_integrated_heats(filename)
```
Think about auto-generating markdown reports, compiling with sphinx or something.

jchodera commented on python/refactored/ITC.py:
```diff
@@ -0,0 +1,191 @@
+        nburn  = 500000 # number of burn-in iterations
+        nthin  = 250 # thinning period
+        
+        mcmc = pymc.MCMC(model, db='ram')
```
`pymc` model should be defined in `models` as part of `TwoComponentBindingModel`.  Perhaps model could have `modeler.getSampler()`?

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+    * time between the beginning of the injection and the beginning of the next injection
+    * filtering period over which data channel is averaged to produce a single measurement of applied power
+
+    EXAMPLES
```
Add docstring test!

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+        # sequence number of injection
+        self.number = number
+        # programmed volume of injection
+        self.volume = volume
```
Consider adding type checking (e.g. compatible with volume units) in constructor?

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+
+        # Store additional data about experiment.
+        parsecline = 11 + self.number_of_injections
+        self.syringe_concentration = float(
```
For future, consider allowing syringe and cell to contain `Solution` kind of objects.

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+        # Process data.
+        nmeasurements = 0
+        injection_labels = list()
+        for (index, line) in enumerate(measurement_lines):
```
This should be cleaned up.

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+            self.injections = self.injections[0:number_of_injections_read]
+            self.number_of_injections = number_of_injections_read
+
+        # DEBUG
```
Add verbosity and/or logging control?  Ask @kyleabeauchamp about logger experiences.

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+
+        return
+
+    def fit_baseline(self):
```
Look into replacing this with a linearly interpolated baseline or maybe Gaussian process from @pgrinaway.
We should figure out how to expose baseline parameters as part of model.  Maybe have `BaselineModel` and `BindingModel` as part of whole model?

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+
+        """
+
+        string = ""
```
Simple way to clean this up is:

```
string = """\
My report stuff here...
%(variablename)s
""" % mydict
```

jchodera commented on python/refactored/experiments.py:
```diff
@@ -0,0 +1,817 @@
+
+        return string
+
+    def write_integrated_heats(self, filename):
```
We should add a corresponding `read_integrated_heats()` for this format that replaces internal integrated heats with those from origin.

jchodera commented on python/refactored/instruments.py:
```diff
@@ -0,0 +1,34 @@
+    def __init__(self):
+        self.V0 = 1.4301 * unit.milliliter  # volume of calorimeter sample cell
+        # Tellinghuisen volume correction for VP-ITC
+        self.V0 = self.V0 - 0.044 * unit.milliliter
```
Expose `V0` and `V0_correction` as an optional arguments.
Add `description` or `name` info?

jchodera commented on python/refactored/instruments.py:
```diff
@@ -0,0 +1,34 @@
+        pass
+
+
+class VPITC(Instrument):
```
Add `ITC200` and `AutoITC200`?

jchodera commented on python/refactored/models.py:
```diff
@@ -0,0 +1,854 @@
+# Rescaling StepMethod for sampling correlated changes in ligand and receptor concentration
+#=============================================================================================
+
+class RescalingStep(pymc.StepMethod):
```
We should have a base `BindingModels` that has this `RescalingStep` as a private thing?
See if this is still necessary?

kyleabeauchamp commented:
Regarding logging, here are some references that seem to strongly advocate for using it:

http://stackoverflow.com/questions/6918493/in-python-why-use-logging-instead-of-print

http://victorlin.me/posts/2012/08/26/good-logging-practice-in-python

http://blog.tplus1.com/blog/2007/09/28/the-python-logging-module-is-much-better-than-print-statements/

jchodera commented on python/refactored/models.py:
```diff
@@ -0,0 +1,854 @@
+#=============================================================================================
+
+
+class BindingModel(object):
```
Move generation of the pymc sampler into a method in this base class: `createSampler()`?

jchodera commented on python/refactored/models.py:
```diff
@@ -0,0 +1,854 @@
+
+        """
+        return exp(-2.0*log_sigma)
+
```
Add specific `createModel()` for this binding model (move from ITC.py).

jchodera commented on python/refactored/models.py:
```diff
@@ -0,0 +1,854 @@
+# MAIN AND TESTS
+#=============================================================================================
+
+if __name__ == "__main__":
```
Get rid of this---move to nose tests?

jchodera commented on python/refactored/report.py:
```diff
@@ -0,0 +1,293 @@
+"""Contains Report class for presenting summary of data in organized fashion."""
```
Replace this by markdown?

jchodera commented on python/refactored/report.py:
```diff
@@ -0,0 +1,293 @@
+        return
+  
+
+def analyze(name, experiment):
```
Should be called something different?

Commit:
Changed from SimTk.unit to pint.

Commit:
Added units.py

Commit:
Some comment and logging fixes.

Commit:
Added a docopt parser.

Commit:
Added model as input option to ITC.py

Commit:
Parser formatting. Minor doctring update.

Commit:
Moved model to mcmc options.

Commit:
Added John's line comments as TODO statements.

Line comments tend to get lost as code changes. These will stay in place and move along with the code. Some IDE's (pycharm, kate) also pick these up and allow summaries as a todo list.

Commit:
Updated parser. Added Instruments.

Commit:
Major parser overhaul.

Commit:
Small fixes.

Commit:
Fix to parser bugs, some unit refactoring.

Commit:
Fixed some unit madness. Misc parser bug fixed.

Commit:
Restructured directories. Working on origin heats parser.

Commit:
Read heats.

Now seems to sample.

Commit:
Bugfixes and old prints.

Commit:
Now actually samples, but crashes randomly.

Commit:
Fix string format to deal with NoneType.

Commit:
Made expected_injection_heats static

Commit:
Units wrapped around expected_injection_heats.

At least one order of magnitude faster. (6000 iterations in 45 seconds instead of 500.)

Commit:
Made nitpic compatible. Fix (some) mcmc result plots.

Commit:
Removed try statement from ITC.py

Commit:
Added seaborn for plots. Removed scommented code.

Commit:
Added sample files to test with.

Commit:
Move output files to workdir.

Commit:
Fix for non-existent workdir.

Commit:
Remove debugging commandline input.

Commit:
Update readme.
